src/servers/src/grpc/flight.rs

```diff
@@ -168,13 +189,70 @@ impl FlightCraft for GreptimeRequestHandler {
         );
         async {
             let output = self.handle_request(request, Default::default()).await?;
-            let stream: Pin<Box<dyn Stream<Item = Result<FlightData, Status>> + Send + Sync>> =
-                to_flight_data_stream(output, TracingContext::from_current_span());
+            let stream = to_flight_data_stream(output, TracingContext::from_current_span());
             Ok(Response::new(stream))
         }
         .trace(span)
         .await
     }
+
+    async fn do_put(
+        &self,
+        request: Request<Streaming<FlightData>>,
+    ) -> TonicResult<Response<TonicStream<PutResult>>> {
+        const MAX_PENDING_RESPONSES: usize = 32;
+        let (tx, rx) = mpsc::channel::<TonicResult<GreptimeResponse>>(MAX_PENDING_RESPONSES);
+
+        let stream = request.into_inner();
+        let stream: PutRecordBatchRequestStream = stream
+            .and_then(|x| future::ready(PutRecordBatchRequest::try_from(x).map_err(Into::into)))
+            .boxed();
+        self.put_record_batches(stream, tx).await;
+
+        let response = ReceiverStream::new(rx)
+            .map_ok(|x| PutResult {
+                app_metadata: bytes::Bytes::from(x.encode_to_vec()),
+            })
+            .boxed();
+        Ok(Response::new(response))
+    }
+}
+
+pub(crate) type PutRecordBatchRequestStream =
+    Pin<Box<dyn Stream<Item = TonicResult<PutRecordBatchRequest>> + Send>>;
+
+pub(crate) enum PutRecordBatchRequest {
+    First((TableName, RawRecordBatch)),
+    Rest(RawRecordBatch),
+}
+
+impl TryFrom<FlightData> for PutRecordBatchRequest {
+    type Error = error::Error;
+
+    fn try_from(value: FlightData) -> Result<Self, Self::Error> {
+        let FlightData {
+            flight_descriptor,
+            data_body,
+            ..
```

Comment:
The `request_id` is in 'app_metadata'; please ensure the PutResult response includes it.

Reply:
See https://github.com/GreptimeTeam/greptimedb-ingester-java/blob/b04b404088f1275c2c08cfa454ad77cf1f704401/ingester-bulk-protocol/src/main/java/io/greptime/Metadata.java#L33